### PR TITLE
Add tests and fixes for store asynchronous behaviors, update typings.

### PR DESCRIPTION
**Pull request by lzhoucs**
**Bug fixes and Tests**

**Description:** 
- Add a simple async storage and unit tests to test store async behaviors.
- Bug fixes.
- Update typings.

**Related Issue:** #30 

Please review this checklist before submitting your PR:
- [x] There is a related issue
- [x] All contributors have signed a CLA
- [ ] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
- [x] The code passes the CI tests
- [x] Unit or Functional tests are included in the PR
- [x] The PR increases or maintains the overall unit test coverage percentage
- [x] The code is ready to be merged

### CODE REVIEW

**Review by kitsonk — APPROVED** [review]
👍 🎃

**Comment by kitsonk** [comment]
There is now a conflict that needs to be resolved.

File: src/store/createStore.ts [hunk 1]
```diff
@@ -47,19 +47,26 @@ export interface StoreFactory extends ComposeFactory<Store<{}, {}, any>, StoreOp
...
+			return state.storage.get(Array.isArray(ids) ? ids : [ ids ]);
+		});
+
 	},
 
 	add(this: Store<{}, {}, any>, items: {}[] | {}, options?: CrudOptions) {
 		const self = this;
-		const storeResultsPromise = instanceStateMap.get(self).storage
-			.add(Array.isArray(items) ? items : [ items ], options);
+		const state = instanceStateMap.get(self);
```
Comment (tomdye) [inline src/store/createStore.ts:66]: should we really be using `self` all over the place?

File: tests/unit/store/createStore.ts [hunk 2]
```diff
@@ -370,5 +380,80 @@ registerSuite({
...
+
+			const start = Date.now();
+			store.add(createData()).then(dfd.callback(function(result: ItemType[]) {
+				const finish = Date.now();
+				assert.isAbove(finish - start, 19 );
+				dfd.resolve();
+			}));
+		},
+		'should complete initial add before subsequent operations'(this: any) {
+			const dfd = this.async(1000);
```
Comment (tomdye) [inline tests/unit/store/createStore.ts:397]: why do you need to use `dfd` here with `this.async`? Can't the tests just return a promise?

File: tests/unit/store/createStore.ts [hunk 3]
```diff
@@ -370,5 +380,80 @@ registerSuite({
...
+			const asyncStorage = createAsyncStorage();
+			const store = createStore({
+				storage: asyncStorage,
+				data: createData()
+			});
+
+			store.get(['1', '2', '3']).then(dfd.callback(function(items: ItemType[]) {
```
Comment (tomdye) [inline tests/unit/store/createStore.ts:404]:
``` javascript
return store.get(['1', '2', '3']).then(items: ItemType[]) {
   assert.deepEqual(items, createData(), 'Didn\'t retrieve items from async add');
});
```

above should return fine, no real need for any of the `dfd = this.async` calls

File: tests/unit/store/createStore.ts [hunk 4]
```diff
@@ -370,5 +380,80 @@ registerSuite({
...
+			const { dfd, store } = getStoreWithAsyncStorage(this, { put: 20, fetch: 10 });
+			store.add(createData());
+			store.fetch().then(function(storeData) {
+				assert.lengthOf(storeData, 0, 'should not have retrieved items');
+			}).then(dfd.resolve);
+		},
+		'async operations should be done in the order specified by the user.'(this: any) {
+			const{ dfd, store } = getStoreWithAsyncStorage(this);
+
+			store.add(createData()).then(function(result) {
```
Comment (tomdye) [inline tests/unit/store/createStore.ts:446]: this looks pretty flakey, are all these `then` calls really needed?

File: tests/unit/store/createStore.ts [hunk 5]
```diff
@@ -370,5 +380,80 @@ registerSuite({
...
+				assert.deepEqual(result, createData(), 'Should have returned all added items');
+			}).then(function() {
+				return store.put(createUpdates()[0]);
```
Comment (tomdye) [inline tests/unit/store/createStore.ts:448]: if `createUpdates` returns the same data every time then why is it not cached?

File: tests/unit/store/createSubcollectionStore.ts [hunk 6]
```diff
@@ -1,8 +1,24 @@
...
-import { createData, ItemType, patches } from '../support/createData';
+import { createData, ItemType, patches, createUpdates } from '../support/createData';
 import { CrudOptions } from '../../../src/store/createStore';
 import createSubcollectionStore from '../../../src/store/createSubcollectionStore';
+import createAsyncStorage from '../support/AsyncStorage';
+
+function getStoreAndDfd(test: any) {
+	const dfd = test.async(1000);
+	const store = createSubcollectionStore<ItemType, CrudOptions>();
+	const Subcollection = store.createSubcollection();
```
Comment (tomdye) [inline tests/unit/store/createSubcollectionStore.ts:11]: is `Subcollection` a class? Not sure why it's capitalised?

File: tests/unit/store/createSubcollectionStore.ts [hunk 7]
```diff
@@ -16,80 +32,71 @@ registerSuite({
...
+			store.fetch().then(function(results) {
+				assert.deepEqual(results, [ data[0] ], 'Didn\'t add item to source store');
+			}).then(dfd.resolve);
+		},
+
+		put(this: any) {
+			const { dfd, store, Subcollection } = getStoreAndDfd(this);
+			const data = createData();
+			Subcollection.put(data[0]);
+			store.fetch().then(function(results) {
```
Comment (tomdye) [inline tests/unit/store/createSubcollectionStore.ts:49]: again, using `dfd` all over the place, this could have been 

``` javascript
return store.fetch().then((results) => {
   assert.deepEqual(results, [ data[0] ], 'Didn\'t put item in source store');
});
```

**Comment by tomdye** [comment]
Didn't comment on every instance of every issue, they repeat throughout the PR

**Review by tomdye — CHANGES_REQUESTED** [review]
Would be nice to see some of these addressed

File: src/store/createStore.ts (hunk 1 above)
Comment (kitsonk) [reply src/store/createStore.ts:66]: No we should be using `this`, not `self`.  But that was not part of this PR.

**Comment by lzhoucs** [comment]
The rebase is done and the conflicts are resolved. However the build for this PR as well the [one just finished for master branch](https://travis-ci.org/dojo/stores/builds/172334823) have been failing, partially because of the typing migration. Will address them now.

Thanks Tom for the comments. I will address them later today.

File: tests/unit/store/createStore.ts (hunk 2 above)
Comment (lzhoucs) [reply tests/unit/store/createStore.ts:397]: For simpler tests like this one, returning a promise is enough. `this.async` used here is for consistency with other places where using `const dfd = this.async(1000);` is necessary. For example, we have a[ bunch of tests](https://github.com/dojo/stores/blob/master/tests/unit/store/mixins/createObservableStoreMixin.ts#L177-L195) that verifies a series of returned value over time. 

Basically this approach allows us to have finer-grained control over test flow.  Also `this.async()` allows us to override the default timeout `30000`.

File: tests/unit/store/createStore.ts (hunk 3 above)
Comment (lzhoucs) [reply tests/unit/store/createStore.ts:404]: I agree. This is for consistency as well. Please see my above comments for details.

File: tests/unit/store/createStore.ts (hunk 4 above)
Comment (lzhoucs) [reply tests/unit/store/createStore.ts:446]: The assertion and the next async operation probably can be combined. I will address this. Thanks!

File: tests/unit/store/createStore.ts (hunk 5 above)
Comment (lzhoucs) [reply tests/unit/store/createStore.ts:448]: Because the store operations could change the cache. `createUpdate()` allows us to always have the original expected test data.

File: tests/unit/store/createSubcollectionStore.ts (hunk 6 above)
Comment (lzhoucs) [reply tests/unit/store/createSubcollectionStore.ts:11]: I actually had the same question. The `getStoreAndDfd()` here [is pulled up](https://github.com/dojo/stores/pull/31/files#diff-781607b689cef84922673aa9bbfce8dbL20) for better reusability and consistency. Maybe @maier49 can let us know.

File: tests/unit/store/createSubcollectionStore.ts (hunk 7 above)
Comment (lzhoucs) [reply tests/unit/store/createSubcollectionStore.ts:49]: I agree. This is for consistency as well. Please see my above comments for details.

**Review by lzhoucs — COMMENTED** [review]
I responded Tom's comments. Will make code changes to address some of them.

File: src/store/createStore.ts (hunk 1 above)
Comment (lzhoucs) [reply src/store/createStore.ts:66]: Yeah. We probably need some cleanup.

File: tests/unit/store/createStore.ts (hunk 2 above)
Comment (tomdye) [reply tests/unit/store/createStore.ts:397]: Right, but why would anything take such a long time? If the process is truly async promises then we should be able to hook onto the `resolve` and `reject` calls to manage the test flow.

File: tests/unit/store/createSubcollectionStore.ts (hunk 6 above)
Comment (maier49) [reply tests/unit/store/createSubcollectionStore.ts:11]: It shouldn't be capitalized. I think that was just a mistake on my part

File: tests/unit/store/createStore.ts (hunk 2 above)
Comment (lzhoucs) [reply tests/unit/store/createStore.ts:397]: I though you suggested to return the `promise` and not to use `dfd.resolve` and `dfd.reject`.

The process is not always async promises. Like [the example I provided above](https://github.com/dojo/stores/blob/master/tests/unit/store/mixins/createObservableStoreMixin.ts#L177-L195) where we have to use `dfd` with `this.async` to test multiple return values. The test goes timeout when assertions fail if we didn't wrap the assertions with try catch. And whenever it happens the default timeout becomes unacceptable. So that is why we put `1000` as the timeout so the tests won't hang for too long in any cases.

Also some of tests you saw in this PR were not added in this PR, I updated the file structure so they are really whitespace changes.

But anyway, I talked with Bradley and we do agree that returning a promise is a cleaner approach. And there's at least 80% of the tests can be refactored for that approach. And we feel like such a change would better be done as a separate PR along with other cleanups after all the [existing PRs](https://github.com/dojo/stores/pulls) are landed first to avoid conflicts.

Comment (tomdye) [reply tests/unit/store/createStore.ts:397]: Ok, can you raise an issue please for this change so that it doesn't get lost? Thanks

Comment (maier49) [reply tests/unit/store/createStore.ts:397]: I opened up an issue for this.
https://github.com/dojo/stores/issues/38

**Review by tomdye — APPROVED** [review]
